README
/*
 * Copyright (c) 2003-2010 Eric Will et al.
 * Rights to this code are documented in doc/LICENSE.
 *
 * General documentation.
 */

Shrike -- a set of services for TSora networks
==============================================

This program is free but copyrighted software; see the LICENSE file for
details.

Information about Shrike may be found at http://shrike.malkier.net/. Development
information and repositories can be found at http://github.com/rakaur/shrike/.

TABLE OF CONTENTS
-----------------
  1. Credits
  2. Presentation
  3. Installation
  4. Command line options
  5. Contact and support

1. CREDITS
----------

While Shrike isn't directly based on any other program it borrows a decent
amount of code from Cygnus and Sentinel.

Information about Cygnus can be found at http://www.habber.net/services/.
Information about Sentinel can be found at http://ircd.botbay.net/sentinel/.

Currently Shrike consists of the following developers:
  rakaur, Eric Will <rakaur@malkier.net>
  dKingston, Michael Rodriguez <dkingston02@gmail.com>

Thanks to testers, contributors, etc:
  sycobuny, Stephen Belcher <sycobuny@malkier.net>
  rintaun, Matt Lanigan <rintaun@projectxero.net>
  Hwy, W. Campbell <wcampbel@botbay.net>
  naki, Adam Walls <naki@coreag.net>
  Rob_X, Mikael <mickek15@hotmail.com>
  madragoran, Mike Campbell <madragoran@avendesora.net>

Special thanks to:
  Hwy, for your continued patience while I badger you endlessly.
  sycobuny, for your continued not-killing-me while I badger your code.
  tanka, for the book I will someday repay you for.
  dbn, for the original Website.

Files that we didn't write most of:
  src/balloc.c, this was stolen from ircd-ratbox.
  src/confp.c, this was stolen from Sentinel (from csircd).
  src/event.c, this was stolen from ircd-ratbox.
  src/match.c, this was stolen from Sentinel (from IRCnet ircd).

2. PRESENTATION
---------------

Shrike is a set of services for TSora-based IRC networks that allows users to
manage their channels in a secure and efficient way and allows operators to
manage various things about their networks.

Shrike currently works with:
  ircd-ratbox 1.3 and 2.x in TS5
  ircd-hybrid 7.0 or later
  ircd-hybrid 6.0 or later
  csircd 1.3 or later

Shrike may *possibly* work with other TSora IRCd's running at least TS3.

3. INSTALLATION
---------------

See the INSTALL file.

4. COMMAND LINE OPTIONS
-----------------------

Normally, Shrike can be run simply by invoking the "shrike" executable.  Shrike
will then use the defaults specified in the configuraiton file and connect
to the specified uplink server. Alternatively, any of the following
command-line options can be specified to change the default behavior:
  -c </path/to/file> - specify the configuration file
  -d                 - start in debugging mode
  -h                 - print the help message and exit
  -n                 - do not fork into the background
  -v                 - print the version information and exit

Shrike also handles the following signals:
  HUP  - force a REHASH
  TERM - force a SHUTDOWN
  USR2 - force a RESTART

Upon starting, Shrike will parse its command-line arguments, open its log file,
parse its configuration file, load its database, connect to the uplink,
and (assuming -n is not specified) detach into the background.  If Shrike
encounters an error while parsing the configuration or database files it will
terminate immediately. Otherwise, it will run until killed or otherwise stopped.

5. CONTACT AND SUPPORT
----------------------

Okay, here's the deal. We enjoy coding this program, and we do it for free. You
didn't pay anything for this, did you? No. Our time and our effort went into
making this program, and now you want us to help you with it, too. Well, that's
fine, but do us a favor: READ THE DOCS. Unlike many other programs we've
included some fairly extensive documentation. The configuration file itself has
a description of every option as well as examples of: how it's used, what it
does, and what happens if you turn it on or off. *Please* don't waste our time
by asking asinine questions which are clearly answered in the documentation.
You'll just make us regret making this program so easy to use that it requires
zero intelligence.

That said, if you've skimmed the docs and can't find your answer feel free to
ask. I've been unable to find the answer I was looking for in many programs
myself so I don't mind. Just don't ask us to set it up for you. It's too easy.

For bug reports, please use this reporting mechanism:
http://github.com/rakaur/shrike/issues

If you're reporting a bug, here's some advice. Be sure to include information
on how to reproduce the problem. If you can't reproduce it, you're likely out
of luck. You can go ahead and report the problem but chances are if we can't
find the cause then we can't fix it. If Shrike crashed (with a core file) be
sure to include a backtrace for us. You can do this by running something along
the lines of "gdb bin/shrike shrike.core" and typing "bt" at the "(gdb)"
prompt. If you do all of these things and still manage to keep your report
short and concise you will be loved.

If your problem requires extensive debugging in a real-time situation we may
ask that you find us on IRC:
  rakaur @ irc.malkier.net
  dKingston @ irc.malkier.net

If you've read this far, congratulations. You are among the few elite people
that actually read documentation. Thank you.


ChangeLog
commit c880166f864968dac8e52d0e91ca62830b5b9a67
Author: andrew12 <goforit7arh@gmail.com>
Date:   Sat Feb 5 00:13:01 2011 -0500

    Fix some markdown stuff.

 README.markdown |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

commit 8c8a36a5822331f17ded31f8b293a1efcda230a0
Author: Eric Will <rakaur@malkier.net>
Date:   Sat Jan 29 00:23:33 2011 -0500

    Fix a core... god this was stupid.

 src/services.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit f4318f5ffd6c34d8144665a16377755554e4e5ac
Author: Eric Will <rakaur@malkier.net>
Date:   Thu Jan 20 10:28:18 2011 -0500

    Update lastlogin time upon logout to prevent premature expirations (JStoker).
    
    The last change fixed expiring logged-in users, but as soon as they logged
    out (if logged in for 30+ days) the next expire run would expire them. Now
    we update the lastlogin time on logout.

 src/node.c     |    4 ++++
 src/services.c |    1 +
 src/shrike.c   |    6 +++---
 3 files changed, 8 insertions(+), 3 deletions(-)

commit 0810709b298cca08edeffc2b267b6fe6f04c1d43
Author: Eric Will <rakaur@malkier.net>
Date:   Wed Dec 8 13:50:07 2010 -0500

    Fix syntax error. Whoops.

 src/services.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 26340cc41b97268aec2d965df0c4251ba3c99fb2
Author: Eric Will <rakaur@malkier.net>
Date:   Wed Dec 8 13:45:44 2010 -0500

    Don't expire logged in usernames.

 src/services.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

commit 9e0b78a262b990b9a06da20caa12dbaff8614a37
Author: Eric Will <rakaur@malkier.net>
Date:   Fri Oct 29 15:42:00 2010 -0400

    Fix a super tricky core where a status mode before param modes exploded.
    
    If you have a status mode, like +o before a param mode, like +l or +k,
    the status mode code would increment the param counter too far and shrike
    would core when trying to set the key or limit. For example, "+okl op key 1"
    would core before this patch. The fix is deceptively difficult.

 src/cmode.c |    8 +-------
 1 files changed, 1 insertions(+), 7 deletions(-)

commit 58eba7605890767ce5b415deab1989933a5702dd
Author: dkingston <dkingston@gir.localdomain>
Date:   Fri Oct 29 13:02:45 2010 -0400

    Do not run shrike as root for security reasons.
    IRC stuff should never be ran as root for this reason, and plus there's absolutely no point to it.

 src/shrike.c |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

commit d40ed06c1ef22c42e8b0390f0361a584b2627d65
Author: Eric Will <rakaur@malkier.net>
Date:   Thu Oct 21 12:29:02 2010 -0400

    Clean up some stuff thanks to clang's static analyzer.

 src/db.c       |    4 ++--
 src/function.c |   17 ++++++-----------
 src/node.c     |    3 +--
 src/services.c |   35 +++++++++++++++++------------------
 4 files changed, 26 insertions(+), 33 deletions(-)

commit 2d4243685d1afc20d4db9d1bcb50ba894d1e1135
Author: Eric Will <rakaur@malkier.net>
Date:   Fri Sep 10 10:36:24 2010 -0400

    Return build settings to dev, etc.

 src/Makefile.in |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

commit 3e5f00f8ab58bd1f2f764fc230b1a1c665dafae8
Author: Eric Will <rakaur@malkier.net>
Date:   Fri Sep 10 10:34:12 2010 -0400

    Fix RECOVER, add UNBANME, and fix a bug in OP.
    
    RECOVER was not updating internal state for modes it changed. Desync!
    RECOVER now works as expected if you're deopped but still in the channel.
    UNBANME was requested, and adds a +e for qualifying hostmasks.
    There was a typo in do_op() that allowed VOPs to use the command! :X

 help/unbanme    |   11 +++++
 src/Makefile.in |    6 +-
 src/help.c      |    2 +
 src/services.c  |  114 +++++++++++++++++++++++++++++++++++++++++++++++++-----
 4 files changed, 119 insertions(+), 14 deletions(-)

commit 4f561e059093ed47cb4994803fe3b94a4f73b98a
Author: Mark Szymanski <mrman208@me.com>
Date:   Thu Sep 9 16:11:27 2010 -0500

    Converted README to Markdown.

 README          |  138 ----------------------------------------------------
 README.markdown |  145 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 145 insertions(+), 138 deletions(-)

commit 15cabadc83413fa66cb96302ab5e9bc3b9c4a70f
Author: Eric Will <rakaur@malkier.net>
Date:   Tue Aug 31 10:08:56 2010 -0400

    Use the one-at-a-time hash by Bob Jenkins (modified for us).

 inc/shrike.h   |   11 ++++-------
 src/function.c |   30 ++++++++++++++++++++----------
 2 files changed, 24 insertions(+), 17 deletions(-)

commit 348c5a92aa0248285ceee3dea9988a6d3740f2eb
Author: Eric Will <rakaur@malkier.net>
Date:   Mon Aug 30 08:48:03 2010 -0400

    Fix a core in conf_rehash().
    
    Cores when conf_rehash() is called and no join_chan is defined in conf.

 src/conf.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit d686b8f4db94b5b688a1a919484129d87c22f6ec
Author: Eric Will <rakaur@malkier.net>
Date:   Thu Aug 26 17:10:24 2010 -0400

    Return version, CFLAGS to dev settings.

 src/Makefile.in |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

commit 0c8c5baf924cc50512d6304dec7c561753b564cf
Author: Eric Will <rakaur@malkier.net>
Date:   Thu Aug 26 17:05:46 2010 -0400

    Update a last few things for a 1.5 release!

 ChangeLog       |   73 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 doc/RELEASE     |    4 ++-
 src/Makefile.in |    6 ++--
 3 files changed, 79 insertions(+), 4 deletions(-)

commit c27393369db7e95dc1dd9bef105e529c073f50c3
Author: Eric Will <rakaur@malkier.net>
Date:   Thu Aug 26 15:23:57 2010 -0400

    Finish cleaning up hash tables! Everything is blazing fast now.
    
    Fixed shash() to use RFC casemapping (ToLower()).
    Got rid of the legacy xHASH() macros.

 inc/extern.h   |    2 +-
 inc/shrike.h   |    4 ++--
 src/conf.c     |    2 +-
 src/function.c |    6 +++---
 src/irc.c      |    2 +-
 src/node.c     |   20 ++++++++++----------
 6 files changed, 18 insertions(+), 18 deletions(-)


src/match.c
/*
 * Copyright (c) 2003-2011 Eric Will <rakaur@malkier.net>
 * Rights to this code are documented in doc/LICENSE.
 *
 * This file contains customized casemapping functions.
 * This code was mostly lifted from ircd 2.10.
 */

#include "../inc/shrike.h"

#define BadPtr(x) (!(x) || (*(x) == '\0'))

int match_mapping = MATCH_RFC1459;

const unsigned char ToLowerTab[] = {
  0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9, 0xa,
  0xb, 0xc, 0xd, 0xe, 0xf, 0x10, 0x11, 0x12, 0x13, 0x14,
  0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d,
  0x1e, 0x1f,
  ' ', '!', '"', '#', '$', '%', '&', 0x27, '(', ')',
  '*', '+', ',', '-', '.', '/',
  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
  ':', ';', '<', '=', '>', '?',
  '@', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i',
  'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's',
  't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '~',
  '_',
  '`', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i',
  'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's',
  't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '~',
  0x7f,
  0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89,
  0x8a, 0x8b, 0x8c, 0x8d, 0x8e, 0x8f,
  0x90, 0x91, 0x92, 0x93, 0x94, 0x95, 0x96, 0x97, 0x98, 0x99,
  0x9a, 0x9b, 0x9c, 0x9d, 0x9e, 0x9f,
  0xa0, 0xa1, 0xa2, 0xa3, 0xa4, 0xa5, 0xa6, 0xa7, 0xa8, 0xa9,
  0xaa, 0xab, 0xac, 0xad, 0xae, 0xaf,
  0xb0, 0xb1, 0xb2, 0xb3, 0xb4, 0xb5, 0xb6, 0xb7, 0xb8, 0xb9,
  0xba, 0xbb, 0xbc, 0xbd, 0xbe, 0xbf,
  0xc0, 0xc1, 0xc2, 0xc3, 0xc4, 0xc5, 0xc6, 0xc7, 0xc8, 0xc9,
  0xca, 0xcb, 0xcc, 0xcd, 0xce, 0xcf,
  0xd0, 0xd1, 0xd2, 0xd3, 0xd4, 0xd5, 0xd6, 0xd7, 0xd8, 0xd9,
  0xda, 0xdb, 0xdc, 0xdd, 0xde, 0xdf,
  0xe0, 0xe1, 0xe2, 0xe3, 0xe4, 0xe5, 0xe6, 0xe7, 0xe8, 0xe9,
  0xea, 0xeb, 0xec, 0xed, 0xee, 0xef,
  0xf0, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7, 0xf8, 0xf9,
  0xfa, 0xfb, 0xfc, 0xfd, 0xfe, 0xff
};

const unsigned char ToUpperTab[] = {
  0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9, 0xa,
  0xb, 0xc, 0xd, 0xe, 0xf, 0x10, 0x11, 0x12, 0x13, 0x14,
  0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d,
  0x1e, 0x1f,
  ' ', '!', '"', '#', '$', '%', '&', 0x27, '(', ')',
  '*', '+', ',', '-', '.', '/',
  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
  ':', ';', '<', '=', '>', '?',
  '@', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I',
  'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
  'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\', ']', '^',
  0x5f,
  '`', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I',
  'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
  'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\', ']', '^',
  0x7f,
  0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89,
  0x8a, 0x8b, 0x8c, 0x8d, 0x8e, 0x8f,
  0x90, 0x91, 0x92, 0x93, 0x94, 0x95, 0x96, 0x97, 0x98, 0x99,
  0x9a, 0x9b, 0x9c, 0x9d, 0x9e, 0x9f,
  0xa0, 0xa1, 0xa2, 0xa3, 0xa4, 0xa5, 0xa6, 0xa7, 0xa8, 0xa9,
  0xaa, 0xab, 0xac, 0xad, 0xae, 0xaf,
  0xb0, 0xb1, 0xb2, 0xb3, 0xb4, 0xb5, 0xb6, 0xb7, 0xb8, 0xb9,
  0xba, 0xbb, 0xbc, 0xbd, 0xbe, 0xbf,
  0xc0, 0xc1, 0xc2, 0xc3, 0xc4, 0xc5, 0xc6, 0xc7, 0xc8, 0xc9,
  0xca, 0xcb, 0xcc, 0xcd, 0xce, 0xcf,
  0xd0, 0xd1, 0xd2, 0xd3, 0xd4, 0xd5, 0xd6, 0xd7, 0xd8, 0xd9,
  0xda, 0xdb, 0xdc, 0xdd, 0xde, 0xdf,
  0xe0, 0xe1, 0xe2, 0xe3, 0xe4, 0xe5, 0xe6, 0xe7, 0xe8, 0xe9,
  0xea, 0xeb, 0xec, 0xed, 0xee, 0xef,
  0xf0, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7, 0xf8, 0xf9,
  0xfa, 0xfb, 0xfc, 0xfd, 0xfe, 0xff
};

int ToLower(int c)
{
  if (match_mapping == MATCH_ASCII)
    return tolower(c);
  /* else */
  return (ToLowerTab[(unsigned char)(c)]);
}

int ToUpper(int c)
{
  if (match_mapping == MATCH_ASCII)
    return toupper(c);
  /* else */
  return (ToUpperTab[(unsigned char)(c)]);
}

void set_match_mapping(int type)
{
  match_mapping = type;
}

#define MAX_ITERATIONS  512
/*
**  Compare if a given string (name) matches the given
**  mask (which can contain wild cards: '*' - match any
**  number of chars, '?' - match any single character.
**
**      return  0, if match
**              1, if no match
*/

/*
** match()
** Iterative matching function, rather than recursive.
** Written by Douglas A Lewis (dalewis@acsu.buffalo.edu)
*/

int match(char *mask, char *name)
{
  u_char *m = (u_char *) mask, *n = (u_char *) name;
  char *ma = mask, *na = name;
  int wild = 0, q = 0, calls = 0;

  if (!*mask)
    return 1;

  /* if the mask is "*", it matches everything */
  if ((*m == '*') && (*(m + 1) == '\0'))
    return 0;

  while (1)
  {
#ifdef  MAX_ITERATIONS
    if (calls++ > MAX_ITERATIONS)
      break;
#endif

    if (*m == '*')
    {
      while (*m == '*')
        m++;
      wild = 1;
      ma = (char *)m;
      na = (char *)n;
    }

    if (!*m)
    {
      if (!*n)
        return 0;
      for (m--; (m > (u_char *) mask) &&
           (*m == '?' || *m == '&' || *m == '#'); m--)
        ;
      if ((m > (u_char *) mask) && (*m == '*') && (m[-1] != '\\'))
        return 0;
      if (!wild)
        return 1;
      m = (u_char *) ma;
      n = (u_char *)++ na;
    }
    else if (!*n)
      return 1;
    if ((*m == '\\') && ((m[1] == '*') || (m[1] == '?') || (m[1] == '&') ||
                         (m[1] == '#') || (m[1] == '%')))
    {
      m++;
      q = 1;
    }
    else
      q = 0;

    if ((ToLower(*m) != ToLower(*n)) &&
        (((*m != '?') && !(*m == '&' && IsAlpha(*n)) &&
          !(*m == '#' && IsDigit(*n)) && !(*m == '%' && IsNon(*n))) || q))
    {
      if (!wild)
        return 1;
      m = (u_char *) ma;
      n = (u_char *)++ na;
    }
    else
    {
      if (*m)
        m++;
      if (*n)
        n++;
    }
  }

  return 1;
}


/*
** collapse a pattern string into minimal components.
** This particular version is "in place", so that it changes the pattern
** which is to be reduced to a "minimal" size.
*/
char *collapse(char *pattern)
{
  char *s = pattern, *s1, *t;

  if (BadPtr(pattern))
    return pattern;
  /*
   * Collapse all \** into \*, \*[?]+\** into \*[?]+
   */
  for (; *s; s++)
    if (*s == '\\')
      if (!*(s + 1))
        break;
      else
        s++;
    else if (*s == '*')
    {
      if (*(t = s1 = s + 1) == '*')
        while (*t == '*')
          t++;
      else if (*t == '?')
        for (t++, s1++; *t == '*' || *t == '?'; t++)
          if (*t == '?')
            *s1++ = *t;
      while ((*s1++ = *t++))
        ;
    }
  return pattern;
}

/*
**  Case insensitive comparison of two null terminated strings.
**
**      returns  0, if s1 equal to s2
**              <0, if s1 lexicographically less than s2
**              >0, if s1 lexicographically greater than s2
*/
int irccasecmp(char *s1, char *s2)
{
  unsigned char *str1 = (unsigned char *)s1;
  unsigned char *str2 = (unsigned char *)s2;
  int res;

  if (!s1 || !s2)
    return -1;

  if (match_mapping == MATCH_ASCII)
    return strcasecmp(s1, s2);

  while ((res = ToUpper(*str1) - ToUpper(*str2)) == 0)
  {
    if (*str1 == '\0')
      return 0;
    str1++;
    str2++;
  }
  return (res);
}

int irccmp(char *s1, char *s2)
{
  unsigned char *str1 = (unsigned char *)s1;
  unsigned char *str2 = (unsigned char *)s2;
  int res;

  if (!s1 || !s2)
    return -1;

  if (match_mapping == MATCH_ASCII)
    return strcmp(s1, s2);

  while ((res = *str1 - *str2) == 0)
  {
    if (*str1 == '\0')
      return 0;
    str1++;
    str2++;
  }
  return (res);
}

int ircncasecmp(char *str1, char *str2, int n)
{
  unsigned char *s1 = (unsigned char *)str1;
  unsigned char *s2 = (unsigned char *)str2;
  int res;

  if (match_mapping == MATCH_ASCII)
    return strncasecmp(str1, str2, n);

  while ((res = ToUpper(*s1) - ToUpper(*s2)) == 0)
  {
    s1++;
    s2++;
    n--;
    if (n == 0 || (*s1 == '\0' && *s2 == '\0'))
      return 0;
  }
  return (res);
}

int ircncmp(char *str1, char *str2, int n)
{
  unsigned char *s1 = (unsigned char *)str1;
  unsigned char *s2 = (unsigned char *)str2;
  int res;

  if (match_mapping == MATCH_ASCII)
    return strncmp(str1, str2, n);

  while ((res = *s1 - *s2) == 0)
  {
    s1++;
    s2++;
    n--;
    if (n == 0 || (*s1 == '\0' && *s2 == '\0'))
      return 0;
  }
  return (res);
}

const unsigned int charattrs[] = {
  /* 0  */ 0,
  /* 1  */ 0,
  /* 2  */ 0,
  /* 3  */ 0,
  /* 4  */ 0,
  /* 5  */ 0,
  /* 6  */ 0,
  /* 7 BEL */ 0,
  /* 8  \b */ 0,
  /* 9  \t */ 0,
  /* 10 \n */ 0,
  /* 11 \v */ 0,
  /* 12 \f */ 0,
  /* 13 \r */ 0,
  /* 14 */ 0,
  /* 15 */ 0,
  /* 16 */ 0,
  /* 17 */ 0,
  /* 18 */ 0,
  /* 19 */ 0,
  /* 20 */ 0,
  /* 21 */ 0,
  /* 22 */ 0,
  /* 23 */ 0,
  /* 24 */ 0,
  /* 25 */ 0,
  /* 26 */ 0,
  /* 27 */ 0,
  /* 28 */ 0,
  /* 29 */ 0,
  /* 30 */ 0,
  /* 31 */ 0,
  /* SP */ 0,
  /* ! */ 0,
  /* " */ 0,
  /* # */ 0,
  /* $ */ 0,
  /* % */ 0,
  /* & */ 0,
  /* ' */ 0,
  /* ( */ 0,
  /* ) */ 0,
  /* * */ 0,
  /* + */ 0,
  /* , */ 0,
  /* - */ 0,
  /* . */ 0,
  /* / */ 0,
  /* 0 */ C_DIGIT,
  /* 1 */ C_DIGIT,
  /* 2 */ C_DIGIT,
  /* 3 */ C_DIGIT,
  /* 4 */ C_DIGIT,
  /* 5 */ C_DIGIT,
  /* 6 */ C_DIGIT,
  /* 7 */ C_DIGIT,
  /* 8 */ C_DIGIT,
  /* 9 */ C_DIGIT,
  /* : */ 0,
  /* ; */ 0,
  /* < */ 0,
  /* = */ 0,
  /* > */ 0,
  /* ? */ 0,
  /* @ */ 0,
  /* A */ C_ALPHA,
  /* B */ C_ALPHA,
  /* C */ C_ALPHA,
  /* D */ C_ALPHA,
  /* E */ C_ALPHA,
  /* F */ C_ALPHA,
  /* G */ C_ALPHA,
  /* H */ C_ALPHA,
  /* I */ C_ALPHA,
  /* J */ C_ALPHA,
  /* K */ C_ALPHA,
  /* L */ C_ALPHA,
  /* M */ C_ALPHA,
  /* N */ C_ALPHA,
  /* O */ C_ALPHA,
  /* P */ C_ALPHA,
  /* Q */ C_ALPHA,
  /* R */ C_ALPHA,
  /* S */ C_ALPHA,
  /* T */ C_ALPHA,
  /* U */ C_ALPHA,
  /* V */ C_ALPHA,
  /* W */ C_ALPHA,
  /* X */ C_ALPHA,
  /* Y */ C_ALPHA,
  /* Z */ C_ALPHA,
  /* [ */ 0,
  /* \ */ 0,
  /* ] */ 0,
  /* ^ */ 0,
  /* _ */ 0,
  /* ` */ 0,
  /* a */ C_ALPHA,
  /* b */ C_ALPHA,
  /* c */ C_ALPHA,
  /* d */ C_ALPHA,
  /* e */ C_ALPHA,
  /* f */ C_ALPHA,
  /* g */ C_ALPHA,
  /* h */ C_ALPHA,
  /* i */ C_ALPHA,
  /* j */ C_ALPHA,
  /* k */ C_ALPHA,
  /* l */ C_ALPHA,
  /* m */ C_ALPHA,
  /* n */ C_ALPHA,
  /* o */ C_ALPHA,
  /* p */ C_ALPHA,
  /* q */ C_ALPHA,
  /* r */ C_ALPHA,
  /* s */ C_ALPHA,
  /* t */ C_ALPHA,
  /* u */ C_ALPHA,
  /* v */ C_ALPHA,
  /* w */ C_ALPHA,
  /* x */ C_ALPHA,
  /* y */ C_ALPHA,
  /* z */ C_ALPHA,
  /* { */ 0,
  /* | */ 0,
  /* } */ 0,
  /* ~ */ 0,
  /* del  */ 0,
  /* 0x80 */ 0,
  /* 0x81 */ 0,
  /* 0x82 */ 0,
  /* 0x83 */ 0,
  /* 0x84 */ 0,
  /* 0x85 */ 0,
  /* 0x86 */ 0,
  /* 0x87 */ 0,
  /* 0x88 */ 0,
  /* 0x89 */ 0,
  /* 0x8A */ 0,
  /* 0x8B */ 0,
  /* 0x8C */ 0,
  /* 0x8D */ 0,
  /* 0x8E */ 0,
  /* 0x8F */ 0,
  /* 0x90 */ 0,
  /* 0x91 */ 0,
  /* 0x92 */ 0,
  /* 0x93 */ 0,
  /* 0x94 */ 0,
  /* 0x95 */ 0,
  /* 0x96 */ 0,
  /* 0x97 */ 0,
  /* 0x98 */ 0,
  /* 0x99 */ 0,
  /* 0x9A */ 0,
  /* 0x9B */ 0,
  /* 0x9C */ 0,
  /* 0x9D */ 0,
  /* 0x9E */ 0,
  /* 0x9F */ 0,
  /* 0xA0 */ 0,
  /* 0xA1 */ 0,
  /* 0xA2 */ 0,
  /* 0xA3 */ 0,
  /* 0xA4 */ 0,
  /* 0xA5 */ 0,
  /* 0xA6 */ 0,
  /* 0xA7 */ 0,
  /* 0xA8 */ 0,
  /* 0xA9 */ 0,
  /* 0xAA */ 0,
  /* 0xAB */ 0,
  /* 0xAC */ 0,
  /* 0xAD */ 0,
  /* 0xAE */ 0,
  /* 0xAF */ 0,
  /* 0xB0 */ 0,
  /* 0xB1 */ 0,
  /* 0xB2 */ 0,
  /* 0xB3 */ 0,
  /* 0xB4 */ 0,
  /* 0xB5 */ 0,
  /* 0xB6 */ 0,
  /* 0xB7 */ 0,
  /* 0xB8 */ 0,
  /* 0xB9 */ 0,
  /* 0xBA */ 0,
  /* 0xBB */ 0,
  /* 0xBC */ 0,
  /* 0xBD */ 0,
  /* 0xBE */ 0,
  /* 0xBF */ 0,
  /* 0xC0 */ 0,
  /* 0xC1 */ 0,
  /* 0xC2 */ 0,
  /* 0xC3 */ 0,
  /* 0xC4 */ 0,
  /* 0xC5 */ 0,
  /* 0xC6 */ 0,
  /* 0xC7 */ 0,
  /* 0xC8 */ 0,
  /* 0xC9 */ 0,
  /* 0xCA */ 0,
  /* 0xCB */ 0,
  /* 0xCC */ 0,
  /* 0xCD */ 0,
  /* 0xCE */ 0,
  /* 0xCF */ 0,
  /* 0xD0 */ 0,
  /* 0xD1 */ 0,
  /* 0xD2 */ 0,
  /* 0xD3 */ 0,
  /* 0xD4 */ 0,
  /* 0xD5 */ 0,
  /* 0xD6 */ 0,
  /* 0xD7 */ 0,
  /* 0xD8 */ 0,
  /* 0xD9 */ 0,
  /* 0xDA */ 0,
  /* 0xDB */ 0,
  /* 0xDC */ 0,
  /* 0xDD */ 0,
  /* 0xDE */ 0,
  /* 0xDF */ 0,
  /* 0xE0 */ 0,
  /* 0xE1 */ 0,
  /* 0xE2 */ 0,
  /* 0xE3 */ 0,
  /* 0xE4 */ 0,
  /* 0xE5 */ 0,
  /* 0xE6 */ 0,
  /* 0xE7 */ 0,
  /* 0xE8 */ 0,
  /* 0xE9 */ 0,
  /* 0xEA */ 0,
  /* 0xEB */ 0,
  /* 0xEC */ 0,
  /* 0xED */ 0,
  /* 0xEE */ 0,
  /* 0xEF */ 0,
  /* 0xF0 */ 0,
  /* 0xF1 */ 0,
  /* 0xF2 */ 0,
  /* 0xF3 */ 0,
  /* 0xF4 */ 0,
  /* 0xF5 */ 0,
  /* 0xF6 */ 0,
  /* 0xF7 */ 0,
  /* 0xF8 */ 0,
  /* 0xF9 */ 0,
  /* 0xFA */ 0,
  /* 0xFB */ 0,
  /* 0xFC */ 0,
  /* 0xFD */ 0,
  /* 0xFE */ 0,
  /* 0xFF */ 0,
};
